reduceSPI/nn.py

```diff
@@ -0,0 +1,890 @@
+    """Map a point in R^N to the tangent space at the identity, i.e.
+    to the Lie Algebra
+    Arg:
+        v = vector in R^N, (..., 3) in our case
+    Return:
+        R = v converted to Lie Algebra element, (3,3) in our case"""
+    # make sure this is a sample from R^3
+    assert v.shape[-1] == 3
+
+    R_x = v.new_tensor([[0., 0., 0.],
+                        [0., 0., -1.],
+                        [0., 1., 0.]])
+
+    R_y = v.new_tensor([[0., 0., 1.],
+                        [0., 0., 0.],
+                        [-1., 0., 0.]])
+
+    R_z = v.new_tensor([[0., -1., 0.],
+                        [1., 0., 0.],
+                        [0., 0., 0.]])
+
+    R = R_x * v[..., 0, None, None] + \
+        R_y * v[..., 1, None, None] + \
+        R_z * v[..., 2, None, None]
+    return R
+
+
+def rodrigues(v):
+    theta = v.norm(p=2, dim=-1, keepdim=True)
+    # normalize K
+    K = map_to_lie_algebra(v / theta)
+
+    I = torch.eye(3, device=v.device, dtype=v.dtype)
+    R = I + torch.sin(theta)[..., None]*K \
+        + (1. - torch.cos(theta))[..., None]*(K@K)
+    return R
+
+
+def add_logvar(rot_mats_no_var, logvar_rotmat):
+    logvar_mat = rodrigues(logvar_rotmat)
+    return rot_mats_no_var @ logvar_mat
+
+
+def convert_rot_mat_to_eul_ang(matrix):
+    y1, y2, z = 0, 0, 0
+    sy = math.sqrt(matrix[0][1]**2+matrix[2][1]**2)
+    if sy > 10**(-6):
+        y1 = math.atan2(matrix[2][1], -matrix[0][1])
+        y2 = math.atan2(matrix[1][2], matrix[1][0])
+        z = math.atan2(sy, matrix[1][1])
+    else:
+        y1 = 0
+        y2 = math.atan2(matrix[1][0], matrix[0][0])
+        z = 0
+    return torch.Tensor([y1, z, y2])
+
+
+def convert_rot_mat_ten_to_wig_mat_ten(matrices, latent_dim):
+    batch_size = matrices.shape[0]
+    euler_tens = torch.zeros((batch_size, 3))
+    for i in range(batch_size):
+        euler_tens[i] = convert_rot_mat_to_eul_ang(matrices[i])
+    wig_matrices = rot_mat_tensor(euler_tens, latent_dim)
+    return wig_matrices
+
+
+SO3 = SpecialOrthogonal(3, point_type="vector")
+
+
+def transform_into_so3(mu, logvar, latent_dim, n_samples=1):
+    n_batch_data, latent_shape = mu.shape
+    sigma = logvar.mul(0.5).exp_()
+    if CUDA:
+        eps = torch.cuda.FloatTensor(
+            n_samples, n_batch_data, latent_shape).normal_()
+    else:
+        eps = torch.FloatTensor(n_samples, n_batch_data,
+                                latent_shape).normal_()
+    eps = eps.reshape(eps.shape[1:])
+    tang_mu = eps*sigma
+    z = mu+tang_mu
+    matrices = rot_mat_tensor(z, latent_dim)
+    return matrices
+
+
+def rot_mat_tensor(tens, latent_dim):
+    batch_size = tens.shape[0]
+    J = Jd[int((latent_dim-1)/2)].to(DEVICE)
+    matrices = torch.Tensor(batch_size, latent_dim, latent_dim).to(DEVICE)
+    for i in range(batch_size):
+        alpha = tens[i][0]
+        beta = tens[i][1]
+        gamma = tens[i][2]
+        matrices[i] = rot_mat(alpha, beta, gamma, int((latent_dim-1)/2), J)
```

Comment:
what does the rot_mat function (imported from another module) do?

Reply:
this function compute the wigner matrix